### PR TITLE
ODC-7210: Create Serverless function using the Import from the Git with Builder Images

### PR DESCRIPTION
**Pull request by Lucifergene**
**Fixes**: 
https://issues.redhat.com/browse/ODC-7210

**Solution Description:**

Added **Serverless Function** Import Strategy to the Git Import Form. Users now can create Serverless Functions directly from the Git Import Form if the Git Repository contains func.yaml file.

**ACs Completed**: 

- [x] Detect func.yaml in Git repo when user enter the Git URL
- [x] Read runtime, builder, buildEnvs and run envs value from the func.yaml. func.yaml
- [x] Proceed with this flow if the builder value is s2i
- [x] Allow users to view and possibly edit the environment variables which are shown in func.yaml. Add the buildEnvs and run envs to the resource 
- [x] if func.yaml is detected change the strategy to Serverless function and do not show builder images in to and  provide a section which allows the user to select the runtime version
- [x] Do not show the Pipeline and Resources section. Default to Serverless Deployment
- [x] Add label function.knative.dev: 'true' to KSVC to visualize the Serverless function in the Topology

**Screen shots / Gifs for design review**: 

https://user-images.githubusercontent.com/47265560/209876416-422b671f-e0ad-4cd7-b35f-701af4ff83fe.mp4

**Screenshot when the New strategy gets selected**

screencapture-localhost-9000-import-ns-avik-2023-01-09-19_58_01

**Screenshot when a Unsupported Runtime mentioned in the Func.yaml**

screencapture-localhost-9000-import-ns-avik-2023-01-16-18_09_55

**Screenshot when a random Runtime mentioned in the Func.yaml**

screencapture-localhost-9000-import-ns-avik-2023-01-16-18_10_15

**With all Strategy Tiles Opened**

screencapture-localhost-9000-import-ns-avik-2023-01-09-20_07_06

**Unit test coverage report**: 

E2E tests added.
Unit tests fixed.

**Test setup:**

1. Install the Serverless Operator
2. Go to the Git Import Add Form.
3. Paste a Git URL having `func.yaml` file. (e.g. https://github.com/Lucifergene/oc-func)
4. You will see the Serverless Function Strategy get selected and the Runtime and version are visible.

**Test Files:**
- `func.yaml` with `runtime` set as `node`: https://bitbucket.org/avikkundu/oc-func
-  `func.yaml` with `runtime` set as `go`: https://bitbucket.org/avikkundu/oc-func-go
-  `func.yaml` with `runtime` set as `god`: https://bitbucket.org/avikkundu/oc-func-god

**Browser conformance**: 

- [x] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge

### CODE REVIEW

**Comment by Lucifergene** [comment]
Propagating docs and px acks from the Epic
/label docs-approved
/label px-approved

**Comment by invincibleJai** [comment]
@Lucifergene frontend job is failing

File: frontend/packages/dev-console/src/components/import/serverlessfunc/func-utils.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import {
+  BuilderImage,
+  NormalizedBuilderImages,
+} from '@console/dev-console/src/utils/imagestream-utils';
```
Comment (invincibleJai) [inline frontend/packages/dev-console/src/components/import/serverlessfunc/func-utils.ts:4]: use relative imports in the same package i.e dev-console here

File: frontend/public/components/build.tsx [hunk 2]
```diff
@@ -121,6 +121,7 @@ export enum BuildStrategyType {
   JenkinsPipeline = 'JenkinsPipeline',
   Source = 'Source',
   Pac = 'Pac',
+  ServerlessFunction = 'ServerlessFunction',
```
Comment (invincibleJai) [inline frontend/public/components/build.tsx:124]: should not have this in public but rather extend it in dev-console/knative as needed

File: frontend/packages/dev-console/src/components/import/import-types.ts [hunk 3]
```diff
@@ -300,6 +300,22 @@ export enum Resources {
...
+export enum Runtime {
+  Node = 'node',
+  NodeJS = 'nodejs',
+  TypeScript = 'typescript',
+  Quarkus = 'quarkus',
+  Go = 'go',
+  Rust = 'rust',
+  SpringBoot = 'springboot',
+  Python = 'python',
+}
```
Comment (vikram-raj) [inline frontend/packages/dev-console/src/components/import/import-types.ts:317]: Serverless function is supported for these runtimes 
`enum Runtime {
  node = 'nodejs',
  nodejs = 'nodejs',
  typescript = 'nodejs',
  quarkus = 'java',
}
`
we do not have support for these runtimes ` ['go', 'rust', 'springboot', 'python']` 
If I add git URL https://github.com/vikram-raj/hello-func-rust for runtime as rust then a strategy is not detected, It should detect the serverless function as func.yaml is present with warning message `“Support for <RUNTIME> is not yet available.”`
image

File: frontend/packages/dev-console/src/components/import/serverlessfunc/FuncSection.tsx [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    return (
+      <FormSection>
+        <Alert
+          className="odc-func-strategy-section__error-alert"
+          isInline
+          variant="danger"
+          title={t('devconsole~Import is not possible.')}
+        >
+          {t(
+            'devconsole~Unsupported Runtime detected. Please update the Repository URL or change the Build Strategy to continue.',
```
Comment (vikram-raj) [inline frontend/packages/dev-console/src/components/import/serverlessfunc/FuncSection.tsx:80]: Not getting this alert for unsupported runtime like rust https://github.com/vikram-raj/hello-func-rust

**Comment by vikram-raj** [comment]
When I enter git URL https://github.com/vikram-raj/hello-func-node it doesn't detect it as a serverless function but func.yaml is present in the repo.
Peek 2023-01-12 13-28

**Review by vikram-raj — APPROVED** [review]
works as expected.

/lgtm

cc: @serenamarie125 
/cc @sanketpathak

File: frontend/packages/dev-console/integration-tests/features/addFlow/create-from-git.feature [hunk 5]
```diff
@@ -246,6 +246,18 @@ Feature: Create Application from git form
               And user is able to see Redirect value is selected in "Insecure traffic"
 
 
+        @regression @odc-7167
+        Scenario: Create Serverless Function from Import From Git Form on Add page: A-06-TC19
+            Given user has installed OpenShift Serverless Operator
+              And user has selected namespace "aut-addflow-git"
+              And user clicks on the Import from Git card on the Add page
+             When user enters Git Repo URL as "https://github.com/Lucifergene/oc-func"
+              And user clicks Create button on Add page
```
Comment (sanketpathak) [inline frontend/packages/dev-console/integration-tests/features/addFlow/create-from-git.feature:255]: Can you also add steps to add the name "oc-func" before creating the application

File: frontend/packages/dev-console/integration-tests/features/addFlow/create-from-git.feature [hunk 6]
```diff
@@ -246,6 +246,18 @@ Feature: Create Application from git form
               And user is able to see Redirect value is selected in "Insecure traffic"
 
 
+        @regression @odc-7167
+        Scenario: Create Serverless Function from Import From Git Form on Add page: A-06-TC19
```
Comment (sanketpathak) [inline frontend/packages/dev-console/integration-tests/features/addFlow/create-from-git.feature:250]:
```suggestion
        Scenario: Create Serverless Function from Import From Git Form on Add page: A-06-TC18
```
Update the following test cases number accordingly

File: frontend/packages/knative-plugin/integration-tests/support/pageObjects/global-po.ts [hunk 7]
```diff
@@ -240,6 +240,7 @@ export const servingPO = {
   routesTab: '[data-test-id="horizontal-link-Routes"]',
   pageDetails: '.pf-c-breadcrumb',
   resourceName: '.co-resource-item__resource-name',
+  resourceIcon: '.co-m-resource-icon',
```
Comment (sanketpathak) [inline frontend/packages/knative-plugin/integration-tests/support/pageObjects/global-po.ts:243]: We already have this in the same file on line number 198

**Comment by sanketpathak** [comment]
@Lucifergene We need the test case in knative package instead of the devconsole package

**Comment by sanketpathak** [comment]
@Lucifergene Do we need the resource type section for the serverless function strategy as it is redundant as per the video?

https://user-images.githubusercontent.com/10252230/213531639-231ad782-c67f-4a9e-9f6f-2cf227c2957d.mp4

**Comment by Lucifergene** [comment]
> @Lucifergene Do we need the resource type section for the serverless function strategy as it is redundant as per the video?
> 
>  Screencast.from.01-20-2023.12.02.38.AM.mp4

We are manually setting the Resource type to `Serverless Deployment`. 
So the idea is if the new Serverless strategy gets selected, then automatically the resource type gets changed to `Serverless Deployment`. That's one of the basic necessities of this new strategy.

**Comment by vikram-raj** [comment]
> @Lucifergene Do we need the resource type section for the serverless function strategy as it is redundant as per the video?
> 
@Lucifergene The resource type section should not be present in the case of a Serverless function strategy. and Resource should be set as Serverless deployment. Thanks @sanketpathak, for pointing this out.

**Comment by sanketpathak** [comment]
Verified on cluster using cluster-bot
[Screencast from 01-20-2023 07:45:11 PM.webm](https://user-images.githubusercontent.com/10252230/213721440-7d9ec7f9-f753-40fc-981d-1ad85b8535d6.webm)
/label qe-approved

**Pull request merged**